sstable: fix two level iterator bug with virtual tables

RaduBerinde opened the pull request:
#### sstable: fix two level iterator bug with virtual tables

Consider the following iterator operations: SeekLT, Prev, Prev
(returns nil), Next.

For a top-level or regular index block, it is possible that
`skipBackward` moves beyond the last index block that intersects the
vtable.  Then, when we do Next, we move back to that block and return
the smallest key inside that block, but this is incorrect if the lower
bound is inside this block.

In normal circumstances, `skipBackward` can't move beyond the last
index block; but it is possible when range key masking is in use: on
the filter causes the index block that intersects the vtable to be
skipped. Then the `rangeKeyMasking` filter is updated, and on the Next
call, the block is no longer skipped.

Note that this can't happen in the other direction - this is because
the index key is an end bound for the indexed block and that makes the
boundary check "tight" in `skipForward` but "loose" in `skipBackward`.

This change makes it so that we use `SeekGE` instead of
`firstInternal` when appropriate.

Fixes #3450

#### sstable: TestVirtualReader cleanup

Some cleanup for this test:
 - use proper directive argument format for bounds
 - remove one-off `twoLevel` argument, replace with already supported
   `block-size` and `index-block-size`
 - use standard formatting for properties
 - rename `citer` dirwective to `compaction-iter`

#### sstable: split virtual reader tests

Split the virtual reader tests into one that is focused on the
properties and one that is focused on iterator correctness. We run the
latter with randomized block sizes.

#### sstable: add virtual reader tests with masking

Add support for controlling masking with an interator command and add
a test cases that reproduces #3450.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/3451)
<!-- Reviewable:end -->

RaduBerinde commented:
There is more subtlety around how this can happen. We have a `rangeKeyMasking` filter. On the last `Prev` call, the filter causes a second-level index block to be skipped. Then the `rangeKeyMasking` filter is updated, and on the `Next` cal, the block is no longer skipped.

jbowens reviewed (commented):
> There is more subtlety around how this can happen. We have a `rangeKeyMasking` filter. On the last `Prev` call, the filter causes a second-level index block to be skipped. Then the `rangeKeyMasking` filter is updated, and on the `Next` cal, the block is no longer skipped.

This makes sense. Range key masking applies within bounds `[s, e)`. When performing forward iteration, the sstable iterator is only responsible for comparing the current key to `e`. It assumes that the current key is ≥ `s`, because it's supposed to be externally guaranteed by the higher iterator stack. By moving the iterator onto a key before the lower virtual bound, we're violating the externally-guaranteed invariant.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-NtqCmRj4mXZ8qBbtkcH:b-c1wzb)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal and @RaduBerinde)

___
*[`sstable/reader_iter_two_lvl.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtqGKp71ThPwKJInGTS:-NtqGKp71ThPwKJInGTT:b-63wc3b) ([raw file](https://github.com/cockroachdb/pebble/blob/ba8c69f8f714bac2f23b872d67a1f8dd3a6e13b6/sstable/reader_iter_two_lvl.go#L909)):*
> ```Go
> 		// i.topLevelIndex.key. If the lower bound cuts through this block, we must
> 		// use SeekGE below.
> 		lowerBoundInsideBlock := i.lower != nil && i.cmp(i.topLevelIndex.ikey.UserKey, i.lower) < 0
> ```

I think we can restrict these key comparisons to when `i.virtualState != nil` because non-virtual bounds are known to the higher iterator stack, and the internal iterator interface indicates that they must translate user operations into the appropriate bounded operation (to prevent exactly these types of comparisons within leaves):

```
// Additionally, the caller needs to ensure that SeekGE/SeekPrefixGE are not
// called with a key > the upper bound, and SeekLT is not called with a key <
// the lower bound. InternalIterator implementations are required to respect
// the iterator bounds, never returning records outside of the bounds with one
// exception: an iterator may generate synthetic RANGEDEL marker records. See
// levelIter.syntheticBoundary for the sole existing example of this behavior.
// Specifically, levelIter can return synthetic keys whose user key is equal to
// the lower/upper bound.
```


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal, @jbowens, and @sumeerbhola)

___
*[`sstable/reader_iter_two_lvl.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtqGKp71ThPwKJInGTS:-NtqZMjs26JBf98nPoaB:b3k0okq) ([raw file](https://github.com/cockroachdb/pebble/blob/ba8c69f8f714bac2f23b872d67a1f8dd3a6e13b6/sstable/reader_iter_two_lvl.go#L909)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we can restrict these key comparisons to when `i.virtualState != nil` because non-virtual bounds are known to the higher iterator stack, and the internal iterator interface indicates that they must translate user operations into the appropriate bounded operation (to prevent exactly these types of comparisons within leaves):

```
// Additionally, the caller needs to ensure that SeekGE/SeekPrefixGE are not
// called with a key > the upper bound, and SeekLT is not called with a key <
// the lower bound. InternalIterator implementations are required to respect
// the iterator bounds, never returning records outside of the bounds with one
// exception: an iterator may generate synthetic RANGEDEL marker records. See
// levelIter.syntheticBoundary for the sole existing example of this behavior.
// Specifically, levelIter can return synthetic keys whose user key is equal to
// the lower/upper bound.
```
</blockquote></details>

That makes sense, I'll add the `virtualState` check. I am working separately on a test for this, I need to extend things so I can mess with the range masking during iteration.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Code LGTM, just waiting on tests and then I think this is gucci

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_two_lvl.go` line 974 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtrNIEz-Sx11oHoGdO2:-NtrNIF-0eFKTYdDWIYq:b98gth0) ([raw file](https://github.com/cockroachdb/pebble/blob/5473ca25dfc475fa9855f2eb25fc2b031dd2db4d/sstable/reader_iter_two_lvl.go#L974)):*
> ```Go
> 			var ikey *InternalKey
> 			var val base.LazyValue
> 			// The keys in the block we just loaded end at topLevelKey. If the upper
> ```

I am not sure this case is needed - at least I can't reproduce a similar failure.

I think the asymmetry here comes from separator keys - when we do `skipBackward`, we stop when `i.cmp(topLevelKey.UserKey, i.lower) < 0`, which is the block *before* the first relevant block (causing issues on the next `skipForward`).

But when we do `skipForward`, we stop when `i.cmp(topLevelKey.UserKey, i.upper) > 0` which is the last relevant block. There might still be an edge case when `UserKey == i.upper`, not sure.


Another difficulty I'm running into w.r.t testing is that I don't think the result is deterministic when we mess with masking.

For example, say the table has keys `a@2 b@2 c@1 d@1`. We set up mask suffix @2, do `First() -> a@2, then Next() -> b@2(), then Next() -> nil`. Then we change the mask suffix to @1 and do `Prev()`. What is this `Prev()` supposed to return? In practice, I believe it could be either `d@1` or `c@1` or `b@2` depending how index blocks line up.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`sstable/reader_iter_two_lvl.go` line 910 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtrajUd9Rto16bPH33q:-NtrajUd9Rto16bPH33r:b94fyxd) ([raw file](https://github.com/cockroachdb/pebble/blob/5473ca25dfc475fa9855f2eb25fc2b031dd2db4d/sstable/reader_iter_two_lvl.go#L910)):*
> ```Go
> 		// use SeekGE below. This is only relevant for virtual tables; iterator
> 		// lower bounds are not enforced during forward iteration.
> 		lowerBoundInsideBlock := i.vState != nil && i.lower != nil && i.cmp(i.topLevelIndex.ikey.UserKey, i.lower) < 0
> ```

I don't think this would work if topLevelIndex is before the first key.. My testing found that we need a similar fix in the single level iterator and doing the same doesn't work.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. I removed the changes to `skipBackward` as I was not able to reproduce a problem there (we'll see if the meta test will run into something).
I added some commits that cleanup and refactor the tests. We now run the iterator tests with many random block sizes. The last test I added breaks without the changes in the two `skipForward` methods.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-NtvNgQ70Y-8n3-hNPS8:bg7igkv)* status: 2 of 6 files reviewed, 11 unresolved discussions (waiting on @jbowens and @RaduBerinde)

___
*[`-- commits` line 8 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv17QN3FYOfw1k_nTR):*
nit: first
___
*[`-- commits` line 13 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv1B7r0O6XQQji7DdV):*
ditto
___
*[`sstable/reader_iter_single_lvl.go` line 1355 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv1ZRM4xDpmhcg1mvM:-Ntv1ZRM4xDpmhcg1mvN:b9vkhgk) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1355)):*
> ```Go
> 	for {
> 		// It is possible that skipBackward went too far and the virtual table lower
> 		// bound cuts through block we are about to load, in which case we must use
> ```

nit: "cuts through" isn't super clear. How about using some more words and saying something like "lower bound is after the first key in the block we are ...". And consider adding a couple of examples of how this can happen, so it's clearer why we are only considering this for virtual tables, and why we need both the conditions in the disjunction.
___
*[`sstable/reader_iter_single_lvl.go` line 1358 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv3fLE6m4CqrjckhXT:-Ntv3fLE6m4CqrjckhXU:b-ncuk5h) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1358)):*
> ```Go
> 		// SeekGE below. The keys in the block we are about to load start right
> 		// after the topLevelIndex key (before we Next).
> 		lowerBoundInsideBlock := i.vState != nil &&
> ```

nit: lowerBoundMayBeInsideBlock?
since this is a necessary condition but not sufficient?
___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtvLpo4CY8Kg6xc-KDw:bev7e6s) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
> ```Go
> 		lowerBoundInsideBlock := i.vState != nil &&
> 			(!i.index.valid() || base.InternalCompare(i.cmp, i.index.ikey, i.vState.lower) < 0)
> ```

singleLevelIterator.skipForward is potentially on the performance sensitive path since it is called whenever we transition blocks in `Next` or `NextPrefix`, and for large key-value pairs that could be very frequent (maybe even after every key-value). In comparison, we shouldn't be transitioning first level index blocks often so not worth worrying about twoLevelIterator.
We've added a key comparison for virtual sstables, which can be expensive for large keys. I wonder whether it is optimizable, in that if `i.data.valid()` before we called `i.data.Next()` in https://github.com/cockroachdb/pebble/blob/9c2da6d20e1178d0bf383f5b848eb1ded43db2de/sstable/reader_iter_single_lvl.go#L1238 then we know we don't need to do this key comparison. So it could be a parameter, say `lowerBoundKnownBeforeBlock` to `skipForward`. Similar before calling `i.data.NextPrefix`.

I am ok with it being a TODO.
___
*[`sstable/reader_iter_single_lvl.go` line 1384 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv9fAk3TYYVM5caaRh:-Ntv2lNq26l98s_p0OzT:bn0j4t) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1384)):*
> ```Go
> 			// is exclusive we use >= below, else we use >.
> 			if i.upper != nil {
> 				cmp := i.cmp(indexKey.UserKey, i.upper)
> ```

Just thinking out aloud: if this block contained a#10, b#20, d#30, we are at risk of the same kind of bug if the last key d#30 does not satisfy the (inclusive or exclusive) upper bound. But in that case the separator will also not satisfy the upper bound.
I think this is a worth a code comment somewhere so that the reader can quickly see why we need this "inside block" logic for `skipForward` and not `skipBackward`.

There is also something slightly subtle about this in that we decided not to load the block, and there is potentially a key in there that violates the upper bound, but that is ok since we will never load this block in the future and start from the end. Because if we did, we would need to check the upper bound.
___
*[`sstable/reader_iter_two_lvl.go` line 907 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvBSPh2I00L1jUUCDa:-NtvBSPh2I00L1jUUCDb:b-h2yl7r) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_two_lvl.go#L907)):*
> ```Go
> 
> 		// It is possible that skipBackward went too far and the virtual table lower
> 		// bound cuts through block we are about to load, in which case we must use
> ```

same nit about "cuts through"
___
*[`sstable/reader_iter_two_lvl.go` line 909 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvBgrT3PvrA_kQPpHV:-NtvBgrT3PvrA_kQPpHW:b-264ntu) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_two_lvl.go#L909)):*
> ```Go
> 		// bound cuts through block we are about to load, in which case we must use
> 		// SeekGE below. The keys in the block we are about to load start right
> 		// after the topLevelIndex key (before we Next).
> ```

right after or equal to the topLevelIndex key ...

(since the same user key can span blocks)


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 2 of 6 files reviewed, 12 unresolved discussions (waiting on @jbowens)

___
*[`sstable/testdata/virtual_reader_iter` line 664 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvPSa16EE3kM7dsDkY:-NtvSIvN2EV7M_Eopsy3:bnvdna0) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/testdata/virtual_reader_iter#L664)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

@jbowens 
 
I am unclear about what real-world situation causes the bug that motivated the PR.
Say the range key masking filter caused block B2 to be irrelevant when doing Prev, and we moved the index to B1 and stopped because nothing in B1 satisfied the lower bound. Then when we do Next, why does the range key masking filter not make block B2 irrelevant again?

Is this because, the SpanChanged to activate the block masking happens after B2 is loaded (from the following comment). That is, we made the decision (I am vaguely remembering some discussions from a long time ago) that block masking does not need to be consistent on whether it masks a block because it is a performance optimization.

```
// The logic to ensure that a mask block-property filter is only applied within
// the bounds of the masking range key is subtle. The interleaving iterator
// guarantees that it never invokes SpanChanged until the point iterator is
// positioned within the range key. During forward iteration, this guarantees
// that any block that a sstable reader might attempt to load contains only keys
// greater than or equal to the range key's lower bound. During backward
// iteration, it provides the analagous guarantee on the range key's upper
// bound.
```
</blockquote></details>

Correct, we do `Prev` which returns nil, then `SpanChanged` changes the filtering, then the block is not longer filtered out when we do `Next`. This corresponds to the second `mask-suffix` directive in the test.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`sstable/testdata/virtual_reader_iter` line 664 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvPSa16EE3kM7dsDkY:-NtvSl3YC-6aendoppSk:bjc72x1) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/testdata/virtual_reader_iter#L664)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Correct, we do `Prev` which returns nil, then `SpanChanged` changes the filtering, then the block is not longer filtered out when we do `Next`. This corresponds to the second `mask-suffix` directive in the test.
</blockquote></details>

IIRC SpanChanged was called with `nil` at that point


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 6 files reviewed, 11 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`-- commits` line 13 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv1B7r0O6XQQji7DdV):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`sstable/reader_iter_single_lvl.go` line 1355 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv1ZRM4xDpmhcg1mvM:-Ntvf2apEhKyEf7vtqMD:b-ya4zpv) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1355)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: "cuts through" isn't super clear. How about using some more words and saying something like "lower bound is after the first key in the block we are ...". And consider adding a couple of examples of how this can happen, so it's clearer why we are only considering this for virtual tables, and why we need both the conditions in the disjunction.
</blockquote></details>

Improved the comment.
___
*[`sstable/reader_iter_single_lvl.go` line 1358 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv3fLE6m4CqrjckhXT:-Ntv_ITZ5YK6ho9ycKNl:b-dmto79) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1358)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: lowerBoundMayBeInsideBlock?
since this is a necessary condition but not sufficient?
</blockquote></details>

Renamed to `useSeek`
___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-Ntvf4Rx07QmS4xPzkRN:b8ap5ju) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

singleLevelIterator.skipForward is potentially on the performance sensitive path since it is called whenever we transition blocks in `Next` or `NextPrefix`, and for large key-value pairs that could be very frequent (maybe even after every key-value). In comparison, we shouldn't be transitioning first level index blocks often so not worth worrying about twoLevelIterator.
We've added a key comparison for virtual sstables, which can be expensive for large keys. I wonder whether it is optimizable, in that if `i.data.valid()` before we called `i.data.Next()` in https://github.com/cockroachdb/pebble/blob/9c2da6d20e1178d0bf383f5b848eb1ded43db2de/sstable/reader_iter_single_lvl.go#L1238 then we know we don't need to do this key comparison. So it could be a parameter, say `lowerBoundKnownBeforeBlock` to `skipForward`. Similar before calling `i.data.NextPrefix`.

I am ok with it being a TODO.
</blockquote></details>

I thought about it and we can elide the check if we have a valid data block, which is the case when we transition blocks. Similarly, in the second-level iterator we can elide it if we have a valid index block.
___
*[`sstable/reader_iter_single_lvl.go` line 1384 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntv9fAk3TYYVM5caaRh:-NtvfPZZ1i3f6riVoqIA:b7rny47) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1384)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Just thinking out aloud: if this block contained a#10, b#20, d#30, we are at risk of the same kind of bug if the last key d#30 does not satisfy the (inclusive or exclusive) upper bound. But in that case the separator will also not satisfy the upper bound.
I think this is a worth a code comment somewhere so that the reader can quickly see why we need this "inside block" logic for `skipForward` and not `skipBackward`.

There is also something slightly subtle about this in that we decided not to load the block, and there is potentially a key in there that violates the upper bound, but that is ok since we will never load this block in the future and start from the end. Because if we did, we would need to check the upper bound.
</blockquote></details>

I added to the comment in `skipForward` explaining why we don't have the symmetrical problem.
___
*[`sstable/reader_iter_two_lvl.go` line 907 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvBSPh2I00L1jUUCDa:-NtvfXC011l4_nkI1LpI:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_two_lvl.go#L907)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same nit about "cuts through"
</blockquote></details>

Done.
___
*[`sstable/reader_iter_two_lvl.go` line 909 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvBgrT3PvrA_kQPpHV:-NtvfXT30OjmYW7COYgL:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_two_lvl.go#L909)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

right after or equal to the topLevelIndex key ...

(since the same user key can span blocks)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtvhCsJ0sV4OklQE-7a:bnir0iq) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I thought about it and we can elide the check if we have a valid data block, which is the case when we transition blocks. Similarly, in the second-level iterator we can elide it if we have a valid index block.
</blockquote></details>

Oh wait, you're right, I'd need to pass through the value of `data.valid()` before that `Next` call.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 6 files reviewed, 11 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtvhyfZ-75c0HScSOJB:b-1mt0yr) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh wait, you're right, I'd need to pass through the value of `data.valid()` before that `Next` call.
</blockquote></details>

Ok, fixed, I used `isDataInvalidated` which means we have no block at all, not that we merely finished iterating it.


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-Ntvvoin-VDgLS3fGWnh:bnfp4nl)* status: 0 of 6 files reviewed, 11 unresolved discussions (waiting on @RaduBerinde and @sumeerbhola)

___
*[`-- commits` line 48 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvqQcA-lKfANmr9wHJ):*
s/interator/iterator/
___
*[`sstable/testdata/virtual_reader_iter` line 664 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvPSa16EE3kM7dsDkY:-NtvueqG4TTW2ZeB_65M:bxldwpa) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/testdata/virtual_reader_iter#L664)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

IIRC SpanChanged was called with `nil` at that point
</blockquote></details>

I forgot about this asymmetry. We require that the iterator observe a key within the range key's bounds (so that we know that the iterator has stepped into the bounds) before we begin applying masking. Asymmetry is a consequence of that decision.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 2 of 6 files at r15, 1 of 4 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 3 of 6 files reviewed, 14 unresolved discussions (waiting on @jbowens and @RaduBerinde)

___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtvyIPe-vhDIqpdPsjh:bpg9he5) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ok, fixed, I used `isDataInvalidated` which means we have no block at all, not that we merely finished iterating it.
</blockquote></details>

I don't quite see why this is correct. And my previous suggestion is probably wrong too.

If block B2 was irrelevant in skipBackward, we could have loaded B1 (which is not masked) and has all keys less than the lower bound. We would do
```
		key, val := i.data.Last()
		if key == nil {
			return nil, base.LazyValue{}
		}
		if i.blockLower != nil && i.cmp(key.UserKey, i.blockLower) < 0 {
			i.exhaustedBounds = -1
			return nil, base.LazyValue{}
		}
```
So i.data is valid.

I think we could make it work by explicitly grabbing `lowerBoundKnownToBeBeforeBlock = i.exhaustedBounds == 0` at the start of `Next`. I am also ok with just merging without optimizing with a TODO.
___
*[`sstable/reader_iter_single_lvl.go` line 1361 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvwoCvEtE6_BruVOyM:-NtvwoCvEtE6_BruVOyN:bmc0t1c) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_single_lvl.go#L1361)):*
> ```Go
> 		// An example of how this can happen:
> 		//
> 		//   Index block 1 - contains keys a@1, c@1
> ```

block 1?
since this is a single level iterator, so only knows of a single index block.
___
*[`sstable/reader_iter_two_lvl.go` line 912 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntw-Z802mIl49mGg13O:-Ntw-Z802mIl49mGg13P:b-5xb5zu) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_two_lvl.go#L912)):*
> ```Go
> 		// An example of how this can happen:
> 		//
> 		//   Top level index block 1 - contains keys a@1, c@1
> ```

This is the second level index block 1, yes?
Which represents a tree that contains these keys.
That is, this comment is talking about the top level index block entry.
___
*[`sstable/reader_iter_two_lvl.go` line 936 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntw0KoUDO6ISkh2CanH:-Ntw0KoUDO6ISkh2CanI:bn2cdhc) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_two_lvl.go#L936)):*
> ```Go
> 		useSeek := i.vState != nil &&
> 			(!i.topLevelIndex.valid() ||
> 				(i.index.isDataInvalidated() && base.InternalCompare(i.cmp, i.topLevelIndex.ikey, i.vState.lower) < 0))
> ```

I think the same issue happens here wrt the all these blocks including the data block being successfully loaded, since `twoLevelIterator.skipBackward` will call `i.singleLevelIterator.lastInternal()`. Also, I worry about getting too clever, and would slightly prefer not to try to optimize this twoLevelIterator case.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 3 of 6 files reviewed, 14 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-Ntw6dcE2ivxDvIxRQAX:b-tqsve8) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't quite see why this is correct. And my previous suggestion is probably wrong too.

If block B2 was irrelevant in skipBackward, we could have loaded B1 (which is not masked) and has all keys less than the lower bound. We would do
```
		key, val := i.data.Last()
		if key == nil {
			return nil, base.LazyValue{}
		}
		if i.blockLower != nil && i.cmp(key.UserKey, i.blockLower) < 0 {
			i.exhaustedBounds = -1
			return nil, base.LazyValue{}
		}
```
So i.data is valid.

I think we could make it work by explicitly grabbing `lowerBoundKnownToBeBeforeBlock = i.exhaustedBounds == 0` at the start of `Next`. I am also ok with just merging without optimizing with a TODO.
</blockquote></details>

Ah.. you are correct. I made one of the keys before the sstable in the test `@2` and the panic triggered. I switched to making it a TODO for now. I will try to find a clean way of doing it separately.
___
*[`sstable/reader_iter_two_lvl.go` line 912 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntw-Z802mIl49mGg13O:-Ntw4ZmS7IGhJPU80t_K:bohwc58) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_two_lvl.go#L912)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is the second level index block 1, yes?
Which represents a tree that contains these keys.
That is, this comment is talking about the top level index block entry.
</blockquote></details>

Yes, it says "top level index block" :)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtwA5_N2r5K9O-fkZfK:b-g2jjms) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah.. you are correct. I made one of the keys before the sstable in the test `@2` and the panic triggered. I switched to making it a TODO for now. I will try to find a clean way of doing it separately.
</blockquote></details>

Also I think if the skipped block was the first block of the physical sst, `exhaustedBounds` would be 0. So I think the correct check is `i.data.isDataInvalidated() || exhaustedBounds was -1`


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 6 files reviewed, 14 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 1361 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvwoCvEtE6_BruVOyM:-NtwA0GEEQzcoKHyXcgc:bt2vr1b) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_single_lvl.go#L1361)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

block 1?
since this is a single level iterator, so only knows of a single index block.
</blockquote></details>

Fixed.
___
*[`sstable/reader_iter_two_lvl.go` line 912 at r18](https://reviewable.io/reviews/cockroachdb/pebble/3451#-Ntw-Z802mIl49mGg13O:-NtwCCQ-9L6-pXsoXJ7t:b-efess6) ([raw file](https://github.com/cockroachdb/pebble/blob/b96fb5789d3b41c47140abd7daea221c2fafc3b4/sstable/reader_iter_two_lvl.go#L912)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, it says "top level index block" :)
</blockquote></details>

Oh sorry, I missed the "entry" part. Yes, updated to `Second-level index block X`.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 0 of 7 files reviewed, 14 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 1360 at r6](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtvLpo4CY8Kg6xc-KDv:-NtwOuYy0iCz6-leZVCx:blu5krz) ([raw file](https://github.com/cockroachdb/pebble/blob/0e8608097efe733fd47ffa38a2c1ccb81dd9748f/sstable/reader_iter_single_lvl.go#L1360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Also I think if the skipped block was the first block of the physical sst, `exhaustedBounds` would be 0. So I think the correct check is `i.data.isDataInvalidated() || exhaustedBounds was -1`
</blockquote></details>

Ok, I found a very nice way to do it in the single level iterator. We already calculate the `blockLower` and set it if the lower bound intersects the block, so we can just use that. PTAL.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 6 files at r24, 1 of 1 files at r25, 1 of 5 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-Ntw_aBp0RdTFf4ZUE9P:bg7igkv)* status: 3 of 7 files reviewed, 15 unresolved discussions (waiting on @jbowens and @RaduBerinde)

___
*[`sstable/reader_iter_single_lvl.go` line 1415 at r28](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtwZljkCNqrYjEXDLwt:-NtwZljkCNqrYjEXDLwu:b-hhr6vm) ([raw file](https://github.com/cockroachdb/pebble/blob/780e5edf308b9849b8cf408f5f3bffe03705256d/sstable/reader_iter_single_lvl.go#L1415)):*
> ```Go
> 		// guarantees wrt an iterator lower bound when we iterate forward. But we
> 		// must never return keys that are not inside the virtual table.
> 		if i.vState != nil && i.blockLower != nil {
> ```

looks good. Our `blockLower` initialization logic seems unnecessarily conservative:
```
	if i.blockLower != nil {
		key, _ := i.data.First()
		if key != nil && i.cmp(i.blockLower, key.UserKey) < 0 {
			// The lower-bound is less than the first key in the block. No need
			// to check the lower-bound again for this block.
			i.blockLower = nil
		}
	}
```
Shouldn't that be `<= 0`? Same for `initBoundsForAlreadyLoadedBlock`.

How about adding a TODO to change it -- I don't want you to have to kickoff another multi-hour meta test run for merging this PR, it it already passes.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-NtxZF9v8HeZKtOpa6V6:bz5mair)* status: 3 of 7 files reviewed, 15 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 1415 at r28](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtwZljkCNqrYjEXDLwt:-NtxZ3EyEhg9ybu9ZEw_:b-hukn7c) ([raw file](https://github.com/cockroachdb/pebble/blob/780e5edf308b9849b8cf408f5f3bffe03705256d/sstable/reader_iter_single_lvl.go#L1415)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

looks good. Our `blockLower` initialization logic seems unnecessarily conservative:
```
	if i.blockLower != nil {
		key, _ := i.data.First()
		if key != nil && i.cmp(i.blockLower, key.UserKey) < 0 {
			// The lower-bound is less than the first key in the block. No need
			// to check the lower-bound again for this block.
			i.blockLower = nil
		}
	}
```
Shouldn't that be `<= 0`? Same for `initBoundsForAlreadyLoadedBlock`.

How about adding a TODO to change it -- I don't want you to have to kickoff another multi-hour meta test run for merging this PR, it it already passes.
</blockquote></details>

Added TODOs


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 6 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451#-:-NtxjSd6ETuiQySO0C10:bg7igkv)* status: 2 of 7 files reviewed, 16 unresolved discussions (waiting on @jbowens and @RaduBerinde)

___
*[`sstable/reader_iter_single_lvl.go` line 302 at r32](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtxieeK6UNBbm3REBFA:-NtxieeK6UNBbm3REBFB:b-cvau5h) ([raw file](https://github.com/cockroachdb/pebble/blob/6f655895be550ed1663a6e67dcf1193b54bc4713/sstable/reader_iter_single_lvl.go#L302)):*
> ```Go
> 	}
> 	i.blockUpper = i.upper
> 	// TODO(radu): this should be >= 0
> ```

`>=0` only if `i.endKeyInclusive` is true.
I think our current code where we have to remember to use both `i.upper` and `i.endKeyInclusive` is asking for bugs. We should (not in this PR of course) try to hide comparisons involving the upper bound behind a method so we don't forget to use both fields. 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3451)* status: 1 of 7 files reviewed, 16 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`sstable/reader_iter_single_lvl.go` line 302 at r32](https://reviewable.io/reviews/cockroachdb/pebble/3451#-NtxieeK6UNBbm3REBFA:-Ntxk1B81qTtUAgO900n:b568vc4) ([raw file](https://github.com/cockroachdb/pebble/blob/6f655895be550ed1663a6e67dcf1193b54bc4713/sstable/reader_iter_single_lvl.go#L302)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`>=0` only if `i.endKeyInclusive` is true.
I think our current code where we have to remember to use both `i.upper` and `i.endKeyInclusive` is asking for bugs. We should (not in this PR of course) try to hide comparisons involving the upper bound behind a method so we don't forget to use both fields.
</blockquote></details>

Yeah, we can use the new `UserKeyBoundary`


<!-- Sent from Reviewable.io -->
